File: task2/src/container/container.cpp
#include "container.h"



Container::~Container() {
    for (int i = 0; i < Size; i++) {
        delete Numbers[i];
        Numbers[i] = nullptr;
    }
}

void Container::CreateContainer(std::ifstream &input) {
    int type = 0;

    while(input >> type) {
        switch (type) {
            case 1:
                Numbers[Size] = new Complex();
                break;
            case 2:
                Numbers[Size] = new Fraction();
                break;
            case 3:
                Numbers[Size] = new Polar();
                break;
        }

        Numbers[Size]->CreateNumber(input);
        ++Size;
    }
}

void Container::CreateRandomContainer(int size) {
    Size = size;

    for (int i = 0; i < size; i++) {
        switch (rand() % 3 + 1) {
            case 1:
                Numbers[i] = new Complex();
                break;
            case 2:
                Numbers[i] = new Fraction();
                break;
            case 3:
                Numbers[i] = new Polar();
                break;
        }
        Numbers[i]->CreateRandomNumber();
    }
}

void Container::PrintContainer(std::ofstream &output) {
    output << "Container size: "<< Size << std::endl;
    for (int i = 0; i < Size; i++) {
        Numbers[i]->PrintNumber(output);
    }
}

void Container::ShakerSort() {
    int left = -1;
    int right = Size;
    int pos = 0;
    bool isChanged = 0;

    do {
        isChanged = 0;

        while (pos + 1 < right) {
            if (Numbers[pos]->ToReal() < Numbers[pos + 1]->ToReal()) {
                std::swap(Numbers[pos], Numbers[pos + 1]);
                isChanged = 1;
            }
            ++pos;
        }
        --pos;
        --right;

        while (pos - 1 > left) {
            if (Numbers[pos - 1]->ToReal() < Numbers[pos]->ToReal()) {
                std::swap(Numbers[pos - 1], Numbers[pos]);
                isChanged = 1;
            }
            --pos;
        }
        ++pos;
        ++left;
    } while (isChanged);
}

File: task3/container.py
from complex import *
from fraction import *
from polar import *

class Container:
    def __init__(self):
        self.store = []

    def ReadStrArray(self, strArray):
        if strArray == ['']:
            return
            
        arrayLen = len(strArray)
        i = 0

        while i < arrayLen:
            key = int(strArray[i])
            i += 1

            if key == 1:
                shape = Complex()
            elif key == 2:
                shape = Fraction()
            elif key == 3:
                shape = Polar()
            
            i = shape.ReadStrArray(strArray, i)
            self.store.append(shape)

    def CreateRandom(self, size):
        for i in range(size):
            key = random.randint(1, 3)

            if key == 1:
                shape = Complex()
            elif key == 2:
                shape = Fraction()
            elif key == 3:
                shape = Polar()
            
            shape.CreateRandom()
            self.store.append(shape)

    def ToString(self):
        res = ""
        for shape in self.store:
            res += shape.ToString() + "\n"
        return res

    def Size(self):
        return len(self.store)

    def ShakerSort(self):
        left = -1
        right = len(self.store)
        pos = 0

        while(True):
            isChanged = False

            while pos + 1 < right:
                if self.store[pos].ToReal() < self.store[pos + 1].ToReal():
                    self.store[pos], self.store[pos + 1] = self.store[pos + 1], self.store[pos]
                    isChanged = True;
                pos += 1

            pos -= 1
            right -= 1

            while pos - 1 > left:
                if self.store[pos - 1].ToReal() < self.store[pos].ToReal():
                    self.store[pos - 1], self.store[pos] = self.store[pos], self.store[pos - 1]
                    isChanged = True;
                pos -= 1
            
            pos += 1
            left += 1

            if isChanged == False:
                break